File: src/main/java/back/ecommerce/cart/service/CartService.java
package back.ecommerce.cart.service;

import static back.ecommerce.exception.ErrorCode.*;
import static java.util.stream.Collectors.*;

import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import back.ecommerce.cart.entity.Cart;
import back.ecommerce.product.entity.Product;
import back.ecommerce.cart.dto.response.AddCartResponse;
import back.ecommerce.cart.dto.response.CartDeleteResponse;
import back.ecommerce.cart.dto.response.CartListResponse;
import back.ecommerce.cart.dto.response.CartProductDto;
import back.ecommerce.cart.dto.response.CartProducts;
import back.ecommerce.exception.CustomException;
import back.ecommerce.exception.ErrorCode;
import back.ecommerce.cart.repository.CartRepository;
import back.ecommerce.product.repository.ProductRepository;
import back.ecommerce.user.repository.UserRepository;
import lombok.RequiredArgsConstructor;

@Service
@RequiredArgsConstructor
public class CartService {

	private final CartRepository cartRepository;
	private final UserRepository userRepository;
	private final ProductRepository productRepository;

	@Transactional
	public AddCartResponse addProduct(String email, long productId, int quantity) {
		validateUserEmail(email);
		Product product = productRepository.findById(productId)
			.orElseThrow(() -> new CustomException(PRODUCT_NOT_FOUND));
		Cart userCart = Cart.create(email, product, quantity);
		return AddCartResponse.create(cartRepository.save(userCart));
	}

	@Transactional(readOnly = true)
	public CartListResponse findCartByUserEmail(String email) {
		validateUserEmail(email);
		CartProducts cartProducts = cartRepository.findByUserEmail(email).stream()
			.map(CartProductDto::create)
			.collect(collectingAndThen(toList(), CartProducts::create));
		return new CartListResponse(email, cartProducts);
	}

	private void validateUserEmail(String email) {
		if (!userRepository.existsByEmail(email)) {
			throw new CustomException(ErrorCode.USER_NOT_FOUND);
		}
	}

	public CartDeleteResponse deleteById(Long cartId, String email) {
		validateUserEmail(email);
		cartRepository.deleteById(cartId);
		return new CartDeleteResponse(email, cartId);
	}
}

File: src/main/java/back/ecommerce/cart/application/CartService.kt
package back.ecommerce.cart.application

import back.ecommerce.cart.dto.response.*
import back.ecommerce.cart.entity.Cart
import back.ecommerce.cart.repository.CartRepository
import back.ecommerce.exception.CustomException
import back.ecommerce.exception.ErrorCode
import back.ecommerce.product.repository.ProductRepository
import back.ecommerce.user.repository.UserRepository
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class CartService(
    private val cartRepository: CartRepository,
    private val userRepository: UserRepository,
    private val productRepository: ProductRepository
) {

    @Transactional
    fun addProduct(email: String, productId: Long, quantity: Int): AddCartResponse {
        validateUserEmail(email)
        val product = productRepository.findById(productId)
            .orElseThrow { CustomException(ErrorCode.PRODUCT_NOT_FOUND) }
        val cart = cartRepository.save(Cart.create(email, product, quantity))
        return AddCartResponse(cart.id, cart.quantity, cart.price)
    }

    @Transactional(readOnly = true)
    fun findCartByUserEmail(email: String): CartListResponse {
        validateUserEmail(email)
        val cartProducts = cartRepository.findByUserEmail(email)
            .map { CartProductDto.create(it) }
            .toList()

        return CartListResponse(email, CartProducts.create(cartProducts))
    }

    private fun validateUserEmail(email: String) {
        if (!userRepository.existsByEmail(email)) {
            throw CustomException(ErrorCode.USER_NOT_FOUND)
        }
    }

    @Transactional
    fun deleteById(cartId: Long, email: String): CartDeleteResponse {
        validateUserEmail(email)
        cartRepository.deleteById(cartId)
        return CartDeleteResponse(email, cartId)
    }
}